crud/common/sharding_key.lua

```diff
@@ -0,0 +1,122 @@
+local fiber = require('fiber')
+local errors = require('errors')
+local vshard = require('vshard')
+
+local const = require('crud.common.const')
+local dev_checks = require('crud.common.dev_checks')
+local cache = require('crud.common.sharding_key_cache')
+local utils = require('crud.common.utils')
+
+local FetchShardingKeyError = errors.new_class('FetchShardingKeyError', {capture_stack = false})
+local GetShardingKeyFieldnosError = errors.new_class('GetShardingKeyFieldnosError',  {capture_stack = false})
```

Comment:
Several thoughts around those errors.

1. Regarding naming. To be honest, 'fieldnos' here confuses me in several ways:
    - 'Fieldno' is the same as 'field №' or 'field ordinal number', right? But there is nothing about field numbers, we perform all transformations on field names.
    - 'Nos' as plural of 'no' (which is '№') doubts me. I don't know whether it is correct way to create a plural noun.
    - 'GetShardingKeyFieldnosError' does not tell me what occurs. I know now, it is about inconsistent database schema: likely invalid sharding key definition. But the name of the error class says nothing for me about that.
2. If I would be a user, I would like to have a high level classification of errors:
    - network error (and it does not matter much what exactly router unable to fetch from storage, at least for a program),
    - schema error (schema inconsistency),
    - incorrect args,
    - timeout error,
    - data inconsistency (maybe),
    - fiber is cancelled.

    In my understanding error classes are mainly for programmatic decisions what to do next after an error. I don't think that we should split error class hierarchy into so specific situations as 'an error during fetching sharding keys', because it does not simplify development of a logic upward (retrying and so on). This information fits better to the error message, which is for a human.

    However it is just my opinion. Maybe I don't see a situation, when one will need to spot exactly this situation. Or I don't see why such little splitting to classes may be suitable for a human.

Reply:
In general:

I believe users pay attention to a errors messages, not it's error code. Sometimes we can show specific error (for example incorrect argument type reported by checks module), sometimes not (for example when we execute remote call using vshard we can show only generic error). What kind of a problem we want to solve here? Make errors more clear for user? Put error messages to a user context? 

In particular:

In the whole patchset three new errors were added:
- `ShardingKeyError:new("Sharding key for space %q is missed in primary index, specify bucket_id", space_name)`
- `FetchShardingKeyError:new('Waiting for sharding key to be fetched is timed out')`
- `GetShardingKeyFieldnosError:new("No such field (%s) in a space format (%s)", field_name, space_name)`

I'm agree regarding error `GetShardingKeyFieldnosError`, I would rename it to something like `WrongShardingConfigurationError`. Other errors I would keep as is.